src/java.base/share/classes/java/util/zip/ZipInputStream.java

```diff
@@ -627,6 +640,91 @@ private void readEnd(ZipEntry e) throws IOException {
         }
     }
 
+    /**
+     * Determine whether the {@link #readEnd(ZipEntry)} method should interpret the
+     * 'compressed size' and 'uncompressed size' fields of the Data Descriptor record
+     * as 64-bit numbers instead of the regular 32-bit numbers.
+     *
+     * Returns true if the LOC has the 'streaming mode' flag set, the 'compressed size' and
+     * 'uncompressed size' are set to the Zip64 magic value 0xFFFFFFFF, and the LOC's extra field
+     * contains a Zip64 field where both 'Original Size' and 'Compressed Size' are set to zero.
+     *
+     * @param extra the LOC extra field to look for a Zip64 field in
+     * @param flag the value of the 'general purpose bit flag' field in the LOC
+     * @param csize the value of the 'compressed size' field in the LOC
+     * @param size  the value of the 'uncompressed size' field in the LOC
+     */
+    private boolean expect64BitDataDescriptor(byte[] extra, int flag, long csize, long size)  {
+        // The LOC's 'general purpose bit flag' 3 must indicate use of a Data Descriptor
+        if ((flag & 8) == 0) {
+            return false;
+        }
+
+        // The LOC's 'compressed size' and 'uncompressed size' must both be marked for Zip64
+        if (csize != ZIP64_MAGICVAL || size != ZIP64_MAGICVAL) {
```

Comment:
The spec for this says different. It says:

>
> 4.4.4 general purpose bit flag:
> ...
>    Bit 3: If this bit is set, the fields crc-32, compressed size and uncompressed size are set to zero in the local header.  The correct values are put in the data descriptor immediately following the compressed data.  

So it expects the value zero for the compressed/uncompressed sizes in the LOC when the data descriptor bit is set.

Reply:
The spec isn't terribly helpful in spelling out what should happen in the case where an entry combines the uses of  data descriptors (mandating that CRC, size and compressed size must be zero) with Zip64 (mandating that size and compressed size must be 0xFFFFFFFF)

My interpretation (based in the InfoZIP implementation) is that in such cases, CRC should be zero, while size and compressed size should be 0xFFFFFFFF, with their counterparts in the Zip64 extra field should set to zero.

Does this make sense?

Reply:
> The spec isn't terribly helpful in spelling out what should happen in the case where an entry combines the uses of data descriptors (mandating that CRC, size and compressed size must be zero) with Zip64 (mandating that size and compressed size must be 0xFFFFFFFF)

I see what you mean. However, given that the data descriptor general bit field is set to indicate that the data descriptor should be honoured, plus the spec stating that the original/compressed sizes in the zip64 extra block aren't mandatory, I think not relying on the zip64 extra block for parsing decisions about a data descriptor content might be better. I think the only role that a zip64 block should play when we are deciding how to parse a data descriptor is whether or not the entry has zip64 extra field set (the header id value of the extra field). Does that sound reasonable?

Reply:
> I think the only role that a zip64 block should play when we are deciding how to parse a data descriptor is whether or not the entry has zip64 extra field set (the header id value of the extra field). Does that sound reasonable?

Are you suggesting that we ONLY look for the presence of a Zip64 (tag 0x1) extended field? Meaning we should ignore the following:

- The values of the 'compressed size' and 'uncompressed size' fields in the LOC header
- The contents/fields in the data block in the Zip64 header

The purpose of this check is to determine wheter the entry "is in the ZIP64 format", to use parlance from `APPNOTE.TXT`. The reason stronger validation was added was to avoid false positives, meaning an entry would be interpreted as "in the Zip64 format", even when that was not the intention of the producer. Some of this validation was on my initiative, but I think maybe some was in response to concerns raised in review.

If we ignore the 'uncompressed size' and 'compresssed size' fields, then if a LOC has both of these set to zero and has a Zip64 extended field, then we'll intepret this as "in the Zip64 format", even though no field is marked using the Zip64 magic value `0xFFFFFFFF`.

I think I would be fine with dropping inspection of the Zip64 contents (as there are probably efforts underway to add validation of LOC Zip64 fields similar to the recently added CEN validation).

But I'm a bit worried that ignoring the LOC values will make us parse some data descriptors using 8 bytes, when the producer did in fact not intend to use the Zip64 format. (The Zip64 field just came through some kind of "pollution" or unintended copying.

A natural way to implement Zip64 parsing is to only look at the extended field if at least one of the relevant LOC headers are `0xFFFFFFFF`, otherwise one can short-circuit and assume there is no Zip64 field.

@LanceAndersen Do you have a cent or two to spare?

Reply:
> @LanceAndersen Do you have a cent or two to spare?

Let me try and dig out from a couple of things and circle back to this again